Python/Product/Analysis/LanguageServer/Settings.cs

```diff
@@ -15,24 +15,62 @@
 // permissions and limitations under the License.
 
 using System;
+using System.Collections.Generic;
 
 namespace Microsoft.PythonTools.Analysis.LanguageServer {
-    internal class LanguageServerSettings {
-        public AutoCompleteSettings autoComplete;
-        public DiagnosticsSettings diagnostics;
+    public sealed class LanguageServerSettings {
+        public class PythonAnalysisOptions {
+            public bool openFilesOnly;
+            public object[] errors = Array.Empty<object>();
+            public object[] warnings = Array.Empty<object>();
+            public object[] information = Array.Empty<object>();
+            public object[] disabled = Array.Empty<object>();
 
-        [Serializable]
-        internal class AutoCompleteSettings {
-            public bool showAdvancedMembers;
-        }
+            private HashSet<object> _errors;
+            private HashSet<object> _warnings;
+            private HashSet<object> _information;
+            private HashSet<object> _disabled;
+
+            public DiagnosticSeverity GetEffectiveSeverity(object code, DiagnosticSeverity defaultSeverity) {
+                Init();
+
+                if (_disabled != null && _disabled.Contains(code)) {
+                    return DiagnosticSeverity.Unspecified;
+                }
+                if (_errors != null && _errors.Contains(code)) {
+                    return DiagnosticSeverity.Error;
+                }
+                if (_warnings != null && _warnings.Contains(code)) {
+                    return DiagnosticSeverity.Warning;
+                }
+                if (_information != null && _information.Contains(code)) {
+                    return DiagnosticSeverity.Information;
+                }
+                return defaultSeverity;
+            }
 
-        [Serializable]
-        internal class DiagnosticsSettings {
-            public DiagnosticSeverity unresolvedImports;
+            public bool Show(DiagnosticSeverity severity) => severity != DiagnosticSeverity.Unspecified;
+
+            private void Init() {
+                if (errors != null && errors.Length > 0) {
+                    _errors = new HashSet<object>(errors);
```

Comment:
I'm not sure I understand this code. Why do we need `errors` field?

Reply:
errors is a string array that comes from VSC (i.e. user stuff). The setting is used to map error code like `unresolved-import` to severity - i.e. we need to know which array the code belongs to (see my PR in vscode-python.
```json
                "python.analysis.errors": {
                    "type": "array",
                    "default": [],
                    "items": {
                        "type": "string"
                    },
                    "description": "List of diagnostics messages to be shown as errors.",
                    "scope": "resource"
                },
```